### ISSUE TITLE
Rule proposal: Require `satisfies` before type assertions (`as`)

### ISSUE THREAD

**Issue by tran-simon** [typescript-eslint/typescript-eslint#10027]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Often in Typescript, you need to use a type assertion using the `as` keyword.

This is dangerous, as you lose all type safety.

A good practice is to use the [satisfies](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#the-satisfies-operator) operator to make sure that the type matches the expected type before transforming it into the new type.

Example:

```typescript
const myFunction = (): MyType=>{
  const v = otherFunctionThatReturnsAComplexType()
  
  // Make sure that `v` is of type `MyComplexType` before casting it to `MyType`
  return v satisfies MyComplexType as MyType
}
```

Without the `satisfies MyComplexType`, if the `otherFunctionThatReturnsAComplexType` function is edited in the future to return a different type, TypeScript would not warn that `return v as MyType` is invalid

### Fail Cases

```TypeScript
v as B;
```

### Pass Cases

```TypeScript
v satisfies A as B;
```

**Comment by kirkwaiblinger** [comment]
Hi @tran-simon! 

Two thoughts -

1. If you don't care whether the `satisfies` and `as` types are related to each other, this is achievable with no-restricted-syntax via `"TSAsExpression[expression.type!=\"TSSatisfiesExpression\"]"` ([playground](https://typescript-eslint.io/play/#ts=5.5.2&showAST=es&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3lAXPTMAligOYDcAUBVNmvhkafAD7woCuAtgEYgyUAKGmigZCaAGaEQdAsRIBKWvUYkWbLr35A&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaaFAF2gEsBjUxAE0OQE9dSBDAD3TAG1JFo0fNEgAaTOGw5kiJDWHdIAFQDKAQWQBRDgAcSyZBXy4eiXfsPGAdKSY7EAQgC8AHWUqVbUhWQAzCijaeiiWuG4AumKSUpAAtiHIbADmiIo6SGwyYABubPAUdF6IYKQAFsW29mAARoi%2BwsWwhrhJYGwldsWZMtDexpDRYAC%2B4ZJDIENAA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)) (if you do this, don't forget to include to also include TSTypeAssertion, the angle bracket syntax)

2. If you _do_ care that the asserted type is wider than the input type, maybe a better way is to ban type assertions altogether and use a generic function for this?
   ```ts
    declare const a: string;

    // just an identity function where we plan to specify the type parameter explicitly
    function widen<T>(x: T): T {
       return x;
    }
    const b = widen<string | number>(a);
    ```
    You could even specify a further constrained type that the input must satisfy. I'll let you play around with 
    ```ts
    function widenIf<TReturn, TSatisfies extends TReturn>(x: TSatisfies): TReturn {
        return x;
    }
    ```
    Note that https://github.com/typescript-eslint/typescript-eslint/issues/7173 also exists to enforce these behaviors on the type assertion syntax directly via the linter.
    
 What do you think? Does one of these address your needs?

**Comment by tran-simon** [comment]
@kirkwaiblinger 

The `widen` function solution is not optiomal, as it requires adding extra JS function calls when all we want is type safety

`no-unsafe-type-assertion` is interesting, but I believe it's for a different use case.

Often, when I use `as`, I know it's unsafe, but it's the only way to go. Enforcing `satisfies` is a small check that can help ensure that `as` was not misused

I've had many cases, especially when using generics, where TypeScript is not smart enough to infer the correct type, so I'm forced to use `as`.

For example: https://stackoverflow.com/questions/68474835/how-to-use-a-switch-statement-to-narrow-down-a-type-using-conditional-types

```typescript
const castFood = <IngredientCategory extends Category>(
  category: IngredientCategory,
  food: Food<any>,
): Food<IngredientCategory> | undefined => {
  switch (category) {
    case "fruit":
      return food satisfies Food<"fruit"> as Food<IngredientCategory>;
    case "grain":
      return food satisfies Food<"grain"> as Food<IngredientCategory>;
    case "meat":
      return food satisfies Food<"meat"> as Food<IngredientCategory>;
  }
  return undefined;
};
```

TypeScript can't tell that when `category` is `fruit`, then `food` is `Food<IngredientCategory>`, so `as` is required.
Using `satisfies` helps making sure that if someone changes the type of `food` the compiler will warn us

**Comment by kirkwaiblinger** [comment]
Understood 👍  so, then, does the first case work as a basic implementation of addressing your use case? Or is something missing?

> 1. If you don't care whether the `satisfies` and `as` types are related to each other, this is achievable with no-restricted-syntax via `"TSAsExpression[expression.type!=\"TSSatisfiesExpression\"]"` ([playground](https://typescript-eslint.io/play/#ts=5.5.2&showAST=es&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3lAXPTMAligOYDcAUBVNmvhkafAD7woCuAtgEYgyUAKGmigZCaAGaEQdAsRIBKWvUYkWbLr35A&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaaFAF2gEsBjUxAE0OQE9dSBDAD3TAG1JFo0fNEgAaTOGw5kiJDWHdIAFQDKAQWQBRDgAcSyZBXy4eiXfsPGAdKSY7EAQgC8AHWUqVbUhWQAzCijaeiiWuG4AumKSUpAAtiHIbADmiIo6SGwyYABubPAUdF6IYKQAFsW29mAARoi%2BwsWwhrhJYGwldsWZMtDexpDRYAC%2B4ZJDIENAA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)) (if you do this, don't forget to include to also include TSTypeAssertion, the angle bracket syntax)

**Comment by kirkwaiblinger** [comment]
FWIW - I'm not sure I fully understand the example, since I don't get type errors when I take away the type assertions completely ([playground](https://www.typescriptlang.org/play/?#code/KYDwDg9gTgLgBDAnmYcDCBDGwDm1FwC8cA5AGZQCuAljCXAD6k5QbUB29TJAtsFiQBQgjtihkMAY1QAxCBAAmAHgCS7FsAXVg7GJmx4oBUNnYKAzuiy58APjgBvQXDjsMfAFxxzMKBxwA3M5wktaGiF5qGlo6emH4wQDaANbAEd6+-gC6Xj5+6kEAvsKiwOJSslS0cCY6FnByikrkVXT2Ti6SEAA20LmZBYLFIrplEtJwAOKsHDUgpvWNyiQsbJy2Hd7UAF7A-fmBQyWj5RMAsvzwtWaWS818Au3B2DworDCUUHsZB0eCSCgGvIFDIoBAePobEZVOovjFdJDwnMFpZEXYiHAonDtAj4kZkXVLC0aHQ4AB+BqtOCRWGaHFxAz4Ak3ZgzTjkqZs6lwC5YIKCLrsHwhDA+JYYmHRelo-HXeoyxC2AAUoUZRhpUtiCoANHAyMCvHcMOxFQBKLzBFx3LF0rV4xWMOCUMzAMgcTREJ4ucwAd1okgAFnAVfbTY5LSLzKgAEQUEnRjwRlxfD5Qdh64ER0JRuDR1YcBNJuApz7p-WKOBZ0Uxh4wQsuBvF4CpsvAysuYrJ5ulp0ut3sTRFAJAA))

**Comment by tran-simon** [comment]
> Understood 👍 so, then, does the first case work as a basic implementation of addressing your use case? Or is something missing?

Yes! Sorry I didn't fully understand your suggestion the first time I read your first comment, but after trying your playground it seems to be exactly what I was looking for!

**Comment by kirkwaiblinger** [comment]
Yay!

I guess it's a question for @typescript-eslint/triage-team whether there's anything else we'd want to do with this issue or close it out.

My recommendation would be that, at the most, we retarget this issue to add the following entry to [the no-restricted-syntax examples](https://typescript-eslint.io/troubleshooting/faqs/general/#how-can-i-ban-specific-language-feature):
```json
{
  "selector": "*:matches(TSAsExpression,TSTypeAssertion)[expression.type!=TSSatisfiesExpression]",
  "messsage": "validate the type explicitly before using a type assertion"
}
```

**Comment by bradzacher** [comment]
I don't quite understand how this adds safety. An `as` assertion isn't some magical operator that allows for anyone to do anything - it is enforced to be a bivariant type operation - meaning that the original type must either be a subtype or a supertype of the asserted type. I.e. you can't do `'' as number`, but you could do `'' as string | number`.

With objects exactly what is a supertype can be a bit fuzzy due to structural typing but there is still a fair bit of safety.

Your examples were very, very vague here - you didn't include any actual types to illustrate your motivation - so it's hard to understand what cases you're trying to guard against.

All that being said - I don't believe that this proposal belongs in this project. It is a very niche ask and the broader community will not find a use in the verbosity of the style.
